substrate/coins/pallet/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+      Self::increase_supply(coin, amount)?;
+
+      let balance = Balance { coin, amount: Amount(amount) };
+      Self::deposit_event(Event::Mint { address: SeraiAddress(at.0), balance });
+      Ok(())
+    }
+
+    /// Burns amount at the given account, errors if not enough funds to burn.
+    pub fn burn_internal(
+      at: &T::AccountId,
+      coin: Coin,
+      amount: SubstrateAmount,
+      instruction: Option<OutInstruction>,
+    ) -> Result<(), Error<T>> {
+      // don't waste time if amount 0.
+      if amount == 0 {
+        return Ok(());
+      }
+
+      // sub amount from account
+      let new_amount =
+        Self::coins(at, coin).unwrap_or(0).checked_sub(amount).ok_or(Error::<T>::NotEnoughFunds)?;
+
+      if new_amount == 0 {
+        Coins::<T>::remove(at, coin);
+      } else {
+        Coins::<T>::set(at, coin, Some(new_amount));
+      }
+
+      // something should have gone horribly wrong if we error on this.
+      if Self::decrease_supply(coin, amount).is_err() {
+        panic!("we tried to burn more assets than what we have");
+      }
+
+      if coin == Coin::Serai {
+        Self::deposit_event(Event::SriBurn { address: SeraiAddress(at.0), amount: Amount(amount) });
+      } else {
+        let balance = Balance { coin, amount: Amount(amount) };
+        // TODO: error for no instruction?
+        Self::deposit_event(Event::Burn {
+          address: SeraiAddress(at.0),
+          balance,
+          instruction: instruction.unwrap(),
+        });
+      }
+      Ok(())
+    }
+
+    /// Transfers coins from `from` to `to`.
+    pub fn transfer_internal(
+      from: &T::AccountId,
+      to: &T::AccountId,
+      coin: Coin,
+      amount: SubstrateAmount,
+    ) -> Result<(), Error<T>> {
+      // don't waste time if amount 0.
+      if amount == 0 {
+        return Ok(());
+      }
+
+      // sub the amount from "from"
+      let from_amount = Self::coins(from, coin)
+        .unwrap_or(0)
+        .checked_sub(amount)
+        .ok_or(Error::<T>::NotEnoughFunds)?;
+
+      // add to "to"
+      let to_amount = Self::coins(to, coin)
+        .unwrap_or(0)
+        .checked_add(amount)
+        .ok_or(Error::<T>::AmountOverflowed)?;
+
+      // save
+      Coins::<T>::set(to, coin, Some(to_amount));
+      if from_amount == 0 {
+        Coins::<T>::remove(from, coin);
+      } else {
+        Coins::<T>::set(from, coin, Some(from_amount));
+      }
+
+      // TODO: update the supply if fees are really being burned or do that
+      // within the fee code.
+
+      let balance = Balance { coin, amount: Amount(amount) };
+      Self::deposit_event(Event::Transfer {
+        from: SeraiAddress(from.0),
+        to: SeraiAddress(to.0),
+        balance,
+      });
+      Ok(())
+    }
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::call_index(0)]
+    #[pallet::weight((0, DispatchClass::Normal))] // TODO
+    pub fn transfer(
+      o: OriginFor<T>,
```

Comment:
This `o` gets compiled into an enum member (`Pallet::Call::transfer { o: OriginFor<T> }`). It'd be polite to call it `origin`-proper.